Issue #9476 - Fix 9.4.51 NullPointerException

sid-evolphin opened the pull request:
#9476 
Added a null check for the Callback parameter in o.e.j.iAbstractConnection.failedCallback(Callback, Throwable) so that the subsequent code-flow is not aborted; while the root cause may be ascertained in case debug logging is enabled.

Further, in the case of a RejectedExecutionException when submitting the failCallback, the Runnable is run instead of directly invoking the Callback, so that exceptions in the Callback do not abort the synchronous code-flow.

sid-evolphin committed:
Fix 9.4.51 NullPointerException #9476

Added a null check for the Callback parameter in o.e.j.iAbstractConnection.failedCallback(Callback, Throwable) so that the subsequent code-flow is not aborted; while the root cause may be ascertained in case debug logging is enabled.

Further, in the case of a RejectedExecutionException when submitting the failCallback, the Runnable is run instead of directly invoking the Callback, so that exceptions in the Callback do not abort the synchronous code-flow.

sid-evolphin committed:
Fixed checkstyle errors

gregw reviewed (changes requested):
I don't think this is the right solution.   It is ignoring a null callback that should never have been created/executed.

Instead, I think the fix should be in `onCompleteSuccess` and `onCompleteFailure` on `org.eclipse.jetty.server.HttpConnection.SendCallback`.    Both of these methods should call `release()` and if they get null back, then do no more.

I can see other problems there (those methods are not truly atomic), but we are not going to entirely fix jetty-9.4.

sid-evolphin committed:
Issue #9476 Fixed null Callback usages in HttpConnection.SendCallback

1. Fixed the usages of `Callback` returned by `o.e.j.s.HttpConnection.SendCallback.release()` to check for `null` before invoking the `failedCallback(Callback, Throwable)` or `succeeded()` methods.

2. Also, added a try-finally block around the Callback invocations in order to ensure that shutdownOutput will get performed irrespective of whether they end up throwing any exception.

sid-evolphin committed:
checkstyle fixes

sid-evolphin committed:
Issue #9476 Partially reverted AbstractConnection based on review feedback

Partially reverted `o.e.j.i.AbstractConnection.failedCallback(Callback, Throwable)` to not have a null check for the `Callback` parameter, based on review feedback from @gregw.

This avoids changing the contract of the method parameter (from `@NonNull` to `@Nullable`).

sid-evolphin commented:
Thank you very much @gregw for the review and comments.

- I have incorporated your suggested changes to the methods in `o.e.j.s.HttpConnection.SendCallback` that use `release()` which may potentially return a `null` `Callback`.
    - They also now have try-finally blocks to ensure that `shutdownOutput()` does get invoked even if there is an exception in the callback.
- I have reverted the null check in `o.e.j.i.AbstractConnection.failedCallback(Callback, Throwable)` to maintain the method parameter contract of `@NonNull`.

Please re-review this PR. Thank you very much.

gregw commented:
Replaced by #9935

sid-evolphin commented:
> Replaced by #9935

Thanks a lot @gregw . I'm glad you are fixing the root cause.
